File: ports/java/src/gui/kernel/BcsInputArbitrator.java
package gui.kernel;

import core.kernel.BcsEvent;
import core.kernel.BcsObject;
import gui.widgets.BcsWidget;
import java.util.concurrent.locks.ReentrantLock;

public class BcsInputArbitrator extends BcsObject {

    public enum DeviceType {
        Pointer,
        Keyboard,
        Touch,
        Tablet,
        NumDeviceTypes
    }

    private static volatile BcsInputArbitrator instance;
    private final int[] deviceCounts = new int[DeviceType.NumDeviceTypes.ordinal()];
    private final ReentrantLock lock = new ReentrantLock();

    private BcsInputArbitrator() {
        super(null);
    }

    public static BcsInputArbitrator getInstance() {
        if (instance == null) {
            synchronized (BcsInputArbitrator.class) {
                if (instance == null) {
                    instance = new BcsInputArbitrator();
                }
            }
        }
        return instance;
    }

    public int deviceCount(DeviceType type) {
        lock.lock();
        try {
            int ordinal = type.ordinal();
            if (ordinal >= 0 && ordinal < deviceCounts.length) {
                return deviceCounts[ordinal];
            }
            return 0;
        } finally {
            lock.unlock();
        }
    }

    public void setDeviceCount(DeviceType type, int count) {
        lock.lock();
        try {
            int ordinal = type.ordinal();
            if (ordinal >= 0 && ordinal < deviceCounts.length) {
                deviceCounts[ordinal] = count;
            }
        } finally {
            lock.unlock();
        }
    }

    public boolean routeEvent(BcsEvent event, BcsWidget target) {
        if (target != null) {
            return target.event(event);
        }
        return false;
    }
}

File: ports/go/src/gui/kernel/bcsinputarbitrator.go
package kernel

import (
	"sync"
	"fmt"
	"github.com/robertpelloni/btk-go/ports/go/src/core/kernel"
)

type DeviceType int

const (
	DeviceTypePointer DeviceType = iota
	DeviceTypeKeyboard
	DeviceTypeTouch
	DeviceTypeTablet
	NumDeviceTypes
)

// BcsInputArbitrator manages multiple input streams and resolves multi-user ownership
type BcsInputArbitrator struct {
	*kernel.BcsObject
	deviceCounts [NumDeviceTypes]int
	mu           sync.RWMutex
}

var inputInstance *BcsInputArbitrator
var inputOnce sync.Once

func GetInputArbitrator() *BcsInputArbitrator {
	inputOnce.Do(func() {
		inputInstance = &BcsInputArbitrator{
			BcsObject: kernel.NewBcsObject(nil),
		}
	})
	return inputInstance
}

func (a *BcsInputArbitrator) DeviceCount(t DeviceType) int {
	a.mu.RLock()
	defer a.mu.RUnlock()
	if t >= 0 && t < NumDeviceTypes {
		return a.deviceCounts[t]
	}
	return 0
}

func (a *BcsInputArbitrator) SetDeviceCount(t DeviceType, count int) {
	a.mu.Lock()
	defer a.mu.Unlock()
	if t >= 0 && t < NumDeviceTypes {
		a.deviceCounts[t] = count
		fmt.Printf("BcsInputArbitrator: Device count updated for %d -> %d\n", t, count)
	}
}

// RouteEvent is where the multi-user routing logic will live
func (a *BcsInputArbitrator) RouteEvent(e kernel.BcsEvent, target *BcsWidget) bool {
	// Abstract mapping of events to widget receivers based on focus tokens
	if target != nil {
		return target.Event(e)
	}
	return false
}
